pkg/relayer/proxy/rings.go

```diff
@@ -0,0 +1,108 @@
+// TODO(@h5law): Move all this logic out into a shared package
```

Comment:
```suggestion
// TODO_BLOCKER(@h5law): Move all this logic out into a shared package to avoid
// the duplication of core business logic between `pkg/relayer/proxy/rings.go`
// and `pkg/appgateserver/rings.go`
```